common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -190,557 +162,149 @@ public final boolean recycle(T o, Handle<T> handle) {
-
-                    if (dst.dropHandle(element)) {
-                        // Drop the object.
-                        continue;
-                    }
-                    element.stack = dst;
-                    dstElems[newDstSize ++] = element;
-                }
-
-                if (srcEnd == LINK_CAPACITY && head.next != null) {
-                    // Add capacity back as the Link is GCed.
-                    this.head.relink(head.next);
-                }
-
-                head.readIndex = srcEnd;
-                if (dst.size == newDstSize) {
-                    return false;
-                }
-                dst.size = newDstSize;
-                return true;
-            } else {
-                // The destination stack is full already.
-                return false;
+        void toAvailable() {
+            int prev = STATE_UPDATER.getAndSet(this, STATE_AVAILABLE);
+            if (prev == STATE_AVAILABLE) {
+                throw new IllegalStateException("Object has been recycled already.");
             }
         }
     }
 
-    private static final class Stack<T> {
-
-        // we keep a queue of per-thread queues, which is appended to once only, each time a new thread other
-        // than the stack owner recycles: when we run out of items in our stack we iterate this collection
-        // to scavenge those that can be reused. this permits us to incur minimal thread synchronisation whilst
-        // still recycling all items.
-        final Recycler<T> parent;
-
-        // We store the Thread in a WeakReference as otherwise we may be the only ones that still hold a strong
-        // Reference to the Thread itself after it died because DefaultHandle will hold a reference to the Stack.
-        //
-        // The biggest issue is if we do not use a WeakReference the Thread may not be able to be collected at all if
-        // the user will store a reference to the DefaultHandle somewhere and never clear this reference (or not clear
-        // it in a timely manner).
-        final WeakReference<Thread> threadRef;
-        final AtomicInteger availableSharedCapacity;
-        private final int maxDelayedQueues;
-
+    private static final class LocalPool<T> {
+        private final Recycler<T> parent;
         private final int maxCapacity;
-        private final int interval;
-        private final int delayedQueueInterval;
-        DefaultHandle<?>[] elements;
-        int size;
-        private int handleRecycleCount;
-        private WeakOrderQueue cursor, prev;
-        private volatile WeakOrderQueue head;
-
-        Stack(Recycler<T> parent, Thread thread, int maxCapacity, int maxSharedCapacityFactor,
-              int interval, int maxDelayedQueues, int delayedQueueInterval) {
+        private final int ratioInterval;
+        private final ArrayDeque<DefaultHandle<T>> pooledHandles;
+        private final ArrayDeque<DefaultHandle<T>> externalInbox; // Must always be synchronised.
```

Comment:
why not using a mpsc unbounded queue from JCTools here?

`claim()` is supposed to be called by the same thread and (and corresponding drain too), while filling it could happen from several threads; it looks an easy way to save using synchronization

Reply:
Good call. This simplifies the code even more. Performance is roughly unchanged in my measurements.

Reply:
Nice one: if the release vs acquire rate isn't that high it shouldn't indeed, but if there are many interleaving release/acquire yep because of the contention on draining from inbox vs releasing to it